src/main/java/com/fastdtw/dtw/WarpPath.java

```diff
@@ -36,58 +35,105 @@ public int size() {
     }
 
     public int minI() {
-        return tsIindexes.get(0).intValue();
+        return tsIindexes.get(tsIindexes.size() - 1);
     }
 
+
     public int minJ() {
-        return tsJindexes.get(0).intValue();
+        return tsJindexes.get(tsJindexes.size() - 1);
     }
 
     public int maxI() {
-        return tsIindexes.get(tsIindexes.size() - 1).intValue();
+        return tsIindexes.get(0);
     }
 
+
     public int maxJ() {
-        return tsJindexes.get(tsJindexes.size() - 1).intValue();
+        return tsJindexes.get(0);
     }
 
     public void addFirst(int i, int j) {
+        tsIindexes.add(i);
+        tsJindexes.add(j);
+    }
+
+    public void addLast(int i, int j) {
         tsIindexes.add(0, i);
         tsJindexes.add(0, j);
     }
 
-    public void addLast(int i, int j) {
-        tsIindexes.add(i);
-        tsJindexes.add(j);
+
+    public List<Integer> getMatchingIndexesForI(int i) {
+        int index = tsIindexes.indexOf(i);

```

Comment:
I understand that access has to be changed but the change in access should be completely internal to the class, right? As a consequence these extra methods seem to be new convenience methods:

```
getMatchingIndexesForI
getMatchingIndexesForJ
invertedCopy
invert
```
I'm fine with new convenience methods, just want to clarify that you need them because they seem to be unrelated to the PR (although obviously using the new data structure).